### ISSUE TITLE
Bug in ROM::SQL::Associations#graph_join with custom dataset name

### ISSUE THREAD

**Issue by nepalez** [rom-rb/rom#218]
Following [my question in gitter](https://gitter.im/rom-rb/chat?at=5547e6f1414c35031ccec430) on the **use of the relation whose name differs from the corresponding table**, I think I've found the root of the problem. And this seems to be a bug.

Using this relation:

```ruby
class Tasks < ROM::Relation[:sql]
  register_as :tasks        # Differs from the dataset's
  dataset     :mygem_tasks
end
```

I've found that when I declare association (`one_to_many` etc.):

```ruby
class Users < ROM::Relation[:sql]

  one_to_many :tasks, key: :user_id

  def with_tasks
    association_join :tasks, select: %i(value)
  end
end
```

I have to use the name of the repository in `one_to_many`, otherwise it won't be recognized by `ROM::SQL::Relation::ClassMethods` finalizer - it [expects the name of the relation, not the dataset](https://github.com/rom-rb/rom-sql/blob/master/lib/rom/sql/relation/class_methods.rb#L34-L39).

The `association_join` expects just the same name for `ROM::SQL::Associations` [to return proper reflection](https://github.com/rom-rb/rom-sql/blob/master/lib/rom/sql/relation/associations.rb#L49).

But at the same time, when `ROM::SQL::Associations` [builds graph](https://github.com/rom-rb/rom-sql/blob/2b42fc7ff3e7ad6c7231cbf47e9b1ecd74937cbc/lib/rom/sql/relation/associations.rb#L96-L99), **it treats the name `:tasks` not as the relation but as the db table**.

I think, the problem is in the method [ROM::SQL::Associations#graph_join](https://github.com/rom-rb/rom-sql/blob/2b42fc7ff3e7ad6c7231cbf47e9b1ecd74937cbc/lib/rom/sql/relation/associations.rb#L48-L60). IMHO it should substitute the name of relation (its `assoc_name` argument) by the name of the dataset.

**Comment by solnic** [comment]
Yes this is definitely a bug. When I added this feature we didn't have a way of setting up a custom dataset name. So, relation name was also dataset name. This must be taken into account now in both association DSL and then `graph_join`.

**Comment by nepalez** [comment]
Obviously, here is two options:

1. Change association DSL so that they use the **table** name, not the **relation** one
2. Alternatively we could change the `#graph_join`, so that it use the **relation** name

I'd prefer the first way because it doesn't require the second relation to be defined.

In such a case a proper use of association would be like this:

```ruby
ROM::SQL.migration do
  change do
    create_table :mygem_tasks do
      foreign_key :user_id, :mygem_users
      # ...
    end
  end
end

ROM::SQL.migration do
  change do
    create_table :mygem_users do
      primary_key :id, null: false
      # ...
    end
  end
end

class Users < ROM::Relation[:sql]
  dataset :mygem_users

  one_to_many :mygem_tasks, key: :user_id

  def with_tasks
    association_join :mygem_tasks, select: %i(title)
  end
end
```
Relation for tasks is not necessary at all.

**Comment by nepalez** [comment]
I'm afraid, there is a deeper problem that appears here.

[As it stated in overview](http://rom-rb.org/introduction/overview/), the commands and relations are expected to be separate.
But the implementaion differs: commands back on *relations*, not "pure" *datasets*.

```ruby
class CreateUser < ROM::Commands::Create[:sql]

  relation :users  # command depends on relation
                   # that mediates its access to datastore

end
```

Following overview, I'd expect the syntax to be something like this:

```ruby
class CreateUser < ROM::Commands::Create[:sql]

  repository :default  # like in definition for the relation
  dataset :mygem_users # uses the dataset (db table etc.), not the relation,
                       # because commands knows nothing about relations

end
```

It is the example that would be consistent to the first option in my previous comment. But, well, I understand it could be too late for such a velleity...

**Comment by solnic** [comment]
Not really, it's not a problem. Commands use relations but each relation exposes its dataset so a command could use dataset directly, in fact rom-sql does that in a couple of places.

**Comment by nepalez** [comment]
Of course, you're right: the dataset is accessible.

IMHO the problem is that a Sequel knows nothing about relations, and with it we HAVE TO use datasets. At the same way in association we refer to relations, which is a bit perplexing.

For example, to fix the bug, I forked the repo and [changed integration spec](https://github.com/nepalez/rom-sql/blob/master/spec/integration/read_spec.rb#L37-L40) by renaming the `tasks` repository to `goals` (the `Task` model to `Goal` etc.). And then [added the small patch](https://github.com/nepalez/rom-sql/blob/master/lib/rom/sql/relation/associations.rb#L59) to `#graph_join` that solves the problem.

Well, it works.

But, as I mentioned above, despite the fact the relation name `goals` is used in associations, I still have to use the table name `tasks` [in grouping](https://github.com/nepalez/rom-sql/blob/master/spec/integration/read_spec.rb#L65) because this is the Sequel's context, not the ROM's.

**Comment by solnic** [comment]
Fixed via rom-sql#25
